kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1288 @@
+            it_const->EquationIdVector(ids, aux_ids, r_current_process_info);
+            for (IndexType i = 0; i < ids.size(); ++i) {
+                if (ids[i] < BaseType::mEquationSystemSize) {
+                #ifdef _OPENMP
+                    omp_set_lock(&BaseType::mlock_array[ids[i]]);
+                #endif
+                    auto &row_indices = indices[ids[i]];
+                    row_indices.insert(ids.begin(), ids.end());
+                #ifdef _OPENMP
+                    omp_unset_lock(&BaseType::mlock_array[ids[i]]);
+                #endif
+                }
+            }
+            for (IndexType i = 0; i < aux_ids.size(); ++i) {
+                if (aux_ids[i] < BaseType::mEquationSystemSize) {
+                #ifdef _OPENMP
+                    omp_set_lock(&BaseType::mlock_array[aux_ids[i]]);
+                #endif
+                    auto &row_indices = indices[aux_ids[i]];
+                    row_indices.insert(aux_ids.begin(), aux_ids.end());
+                #ifdef _OPENMP
+                    omp_unset_lock(&BaseType::mlock_array[aux_ids[i]]);
+                #endif
+                }
+            }
+        }
+
+        // Count the row sizes
+        SizeType nnz = 0;
+        for (IndexType i = 0; i < indices.size(); ++i)
+            nnz += indices[i].size();
+
+        rA = CompressedMatrixType(indices.size(), indices.size(), nnz);
+
+        double *Avalues = rA.value_data().begin();
+        IndexType *Arow_indices = rA.index1_data().begin();
+        IndexType *Acol_indices = rA.index2_data().begin();
+
+        // Filling the index1 vector - DO NOT MAKE PARALLEL THE FOLLOWING LOOP!
+        Arow_indices[0] = 0;
+        for (int i = 0; i < static_cast<int>(rA.size1()); i++)
+            Arow_indices[i + 1] = Arow_indices[i] + indices[i].size();
+
+        #pragma omp parallel for
+        for (int i = 0; i < static_cast<int>(rA.size1()); ++i) {
+            const IndexType row_begin = Arow_indices[i];
+            const IndexType row_end = Arow_indices[i + 1];
+            IndexType k = row_begin;
+            for (auto it = indices[i].begin(); it != indices[i].end(); ++it) {
+                Acol_indices[k] = *it;
+                Avalues[k] = 0.0;
+                k++;
+            }
+
+            indices[i].clear(); //deallocating the memory
+
+            std::sort(&Acol_indices[row_begin], &Acol_indices[row_end]);
+        }
+
+        rA.set_filled(indices.size() + 1, nnz);
+
+        Timer::Stop("MatrixStructure");
+    }
+
+    /**
+     * @brief This function is exactly same as the ConstructMatrixStructure() function in base class except that the function has the call to ApplyConstraints function call once the element and conditions compute their equation slave_ids
+     * @param pScheme The pointer to the integration scheme
+     * @param rT The global relation matrix
+     * @param rModelPart The model part to compute
+    */
+    virtual void ConstructRelationMatrixStructure(
```

Comment:
why don't you do the exact same as for the ConstructMatrixStructure?

Reply:
Because the structure is quite different (rectangular matrix)